numpy/_core/tests/test_multiprocessing.py

```diff
@@ -0,0 +1,51 @@
+import multiprocessing
+from multiprocessing import shared_memory, Process
+
+import pytest
+import numpy as np
+
+pytestmark = pytest.mark.thread_unsafe(
+    reason="tests in this module are explicitly multi-processed"
+)
+
+def bool_array_writer(barrier, shm_name, n):
+    # writer routine for test_read_write_bool_array
+    import time
+    shm = shared_memory.SharedMemory(name=shm_name)
+    arr = np.ndarray(n, dtype=np.bool_, buffer=shm.buf)
+    barrier.wait()
+    for i in range(n):
+        arr[i] = True
+        time.sleep(0.00001)
+
+def bool_array_reader(barrier, shm_name, n):
+    # reader routine for test_read_write_bool_array
+    shm = shared_memory.SharedMemory(name=shm_name)
+    arr = np.ndarray(n, dtype=np.bool_, buffer=shm.buf)
+    barrier.wait()
+    for i in range(n):
+        while not arr[i]: pass
+
+def test_read_write_bool_array():
+    # See: gh-30389
+    #
+    # Prior to Python 3.13, boolean scalar singletons (np.True / np.False) were
+    # regular reference-counted objects. Due to the double evaluation in
+    # PyArrayScalar_RETURN_BOOL_FROM_LONG, concurrent reads and writes of a
+    # boolean array could corrupt their refcounts, potentially causing a crash
+    # (e.g., `free(): invalid pointer`).
+    #
+    # This test creates a multi-process race between a writer and a reader to
+    # ensure that NumPy does not exhibit such failures.
+    barrier = multiprocessing.Barrier(2)
+    n = 10000
+    shm = shared_memory.SharedMemory(create=True, size=n)
+    writer = Process(target=bool_array_writer, args=(barrier, shm.name, n))
+    reader = Process(target=bool_array_reader, args=(barrier, shm.name, n))
+    writer.start()
+    reader.start()
+    writer.join()
+    reader.join()
```

Comment:
I'd probably write this using `ProcessPoolExecutor` to avoid this boilerplate and join the processes by construction with a `with` statement.